onnxruntime/server/executor.cc

```diff
@@ -122,6 +126,12 @@ protobufutil::Status Executor::Predict(const std::string& model_name,
     return GenerateProtobufStatus(status, "Run() failed: " + status.ToString());
   }
 
+  // Manually free all memory in NameMLValMap
```

Comment:
why did you have to do this? MLValue's should be self-destructible as they hold a shared_ptr to the underlying memory.

Reply:
Not always.

Reply:
MLValue has a refcount to the underlying C++ object, which may be a Tensor. However, Tensor may not own its buffer.

Reply:
@tmccrmck can you create the tensor with the deleter so that we don't have to rely on explicit deletes and rely on RAII instead?

Reply:
Tensors never own their own buffer but we do give the MLValue's shared_ptr a deleter when initializing the MLVal [here](https://github.com/microsoft/onnxruntime/blob/c2b412f7be037d309eb92ae58deba1ae43321a5c/onnxruntime/core/framework/tensorprotoutils.cc#L473).

The shared_ptr's deleter comes from [here](https://github.com/microsoft/onnxruntime/blob/c2b412f7be037d309eb92ae58deba1ae43321a5c/onnxruntime/core/framework/data_types.cc#L341) so it does get a deleter. I'm currently seeing that when the MLValue's dtor get's called the ref count is above 1 (3 for this test) so I'm assuming that the ref count gets incremented on a copy somewhere like SequentialExecutor or ExecutionFrame. Not sure.

Reply:
For objects outside of onnxruntime engine, my suggestion is:
1. Don't use allocator
2. Allocate the buffer by yourself, manage its lifecycle by yourself
3. As there is no allocator, Tensor doesn't own his buffer

Reply:
OK as we discussed offline we are currently doing all three of those points.

Reply:
Thanks.

Reply:
Would be nice if this could be done using RAII though.

Reply:
Like the OrtValueArray class and HeapBuffer class in onnx_test_runner